src/useObservable.ts

```diff
@@ -1,98 +1,69 @@
+  /**
+   * Ensures that the initialValue is always up-to-date in case the observable changes.
+   */
   useEffect(() => {
-    const store = getOrCreateStore(observable, getValue(initialValue))
-    if (shouldRestoreSubscriptionRef.current) {
-      if (store.subscription.closed) {
-        store.subscription = store.observable.subscribe()
+    initialValueRef.current = getValue(initialValue) as UnboxObservable<ObservableType>
+  }, [initialValue])
+
+  const store = useMemo(() => {
+    if (!cache.has(observable)) {
+      const entry: Partial<CacheRecord<UnboxObservable<ObservableType>>> = {
+        snapshot: initialValueRef.current,
       }
-      shouldRestoreSubscriptionRef.current = false
+      entry.observable = observable.pipe(
+        shareReplay({refCount: true, bufferSize: 1}),
+        tap((value) => (entry.snapshot = value)),
+      )
+
+      // Eagerly subscribe to sync set `entry.currentValue` to what the observable returns, and keep the observable alive until the component unmounts.
+      entry.subscription = entry.observable.subscribe()
+
+      cache.set(observable, entry as CacheRecord<UnboxObservable<ObservableType>>)
+    }
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const instance = cache.get(observable)!
+    if (instance.subscription.closed) {
+      instance.subscription = instance.observable.subscribe()
     }
 
-    return () => {
-      // React StrictMode will call effects as `setup + teardown + setup` thus we can't trust this callback as "react is about to unmount"
-      // Tracking this ref lets us set the subscription back up on the next `setup` call if needed, and if it really did unmounted then all is well
-      shouldRestoreSubscriptionRef.current = !store.subscription.closed
-      store.subscription.unsubscribe()
+    return {
+      subscribe: (onStoreChange: () => void) => {
+        const subscription = instance.observable.subscribe(onStoreChange)
+        instance.subscription.unsubscribe()
+        return () => subscription.unsubscribe()
+      },
+      getSnapshot: () => instance.snapshot,
     }
   }, [observable])
 
-  return useSyncExternalStore(subscribe, getSnapshot)
+  return useSyncExternalStore<UnboxObservable<ObservableType>>(store.subscribe, store.getSnapshot)
 }
 
-export function useMemoObservable<T>(
-  observableOrFactory: Observable<T> | (() => Observable<T>),
-  deps: DependencyList,
-): T | undefined
-export function useMemoObservable<T>(
-  observableOrFactory: Observable<T> | (() => Observable<T>),
-  deps: DependencyList,
-  initialValue: T | (() => T),
-): T
-export function useMemoObservable<T>(
-  observableOrFactory: Observable<T> | (() => Observable<T>),
-  deps: DependencyList,
-  initialValue?: T | (() => T),
-) {
-  return useObservable(
-    useMemo(() => getValue(observableOrFactory), deps),
-    initialValue,
-  )
-}
+/** @internal */
+export type UnboxObservable<T> = T extends Observable<infer U> ? U : never
```

Comment:
There's an [`ObservedValueOf`](https://rxjs.dev/api/index/type-alias/ObservedValueOf) type exported from rxjs that captures a lot more value types than just observables.

It probably makes sense for us to keep it narrow for now though, but at least adding a separate condition for `InteropObservable` here can potentially save us some compatibility issues later on.

Reply:
I refactored to `ObservedValueOf` and it worked. I couldn't get `InteropObservable` to work well as it caused type errors in the setup that adds `shareReplay` and such 🤔